hk_adc/hk_adc.c

```diff
@@ -0,0 +1,530 @@
+    status = MSS_I2C_wait_complete(&g_mss_i2c0, MSS_I2C_NO_TIMEOUT);
+
+    if (status == MSS_I2C_SUCCESS)
+    {
+        err = HK_ADC_NO_ERR;
+        return err;
+    }
+
+    err = HK_ADC_ERR_WRITE_FAILED;
+    return err;
+}
+
+
+
+
+/**
+ * @brief HK ADC register read function
+ *
+ * This function performs register read operation. Usually, the value to be
+ * read is a 16-bit value as all the registers accessible for a read
+ * operation in HK ADC are 16-bit registers.
+ *
+ * @param reg[in]              Variable of enumeration type #hk_adc_register_t
+ *                             (defined in hk_adc.h) indicating which register
+ *                             is to be read.
+ * @param read_buffer[in]      pointer to a uint16_t variable to store the result
+ *                             in.
+ * @return HK_ADC_ERR_READ_FAILED if any error occurs otherwise HK_ADC_NO_ERR.
+ *
+ */
+int hk_adc_reg_read(hk_adc_register_t reg, uint16_t *read_buffer)
+{
+    uint8_t send_buffer = 0xff;
+    uint8_t err = HK_ADC_ERR_READ_FAILED;
+
+    // assign register address pointer
+    switch(reg)
+    {
+        case HK_ADC_REG_CONVERSION:
+            send_buffer = HK_ADC_REG_PTR_CONVERSION;
+            break;
+
+        case HK_ADC_REG_CONFIG:
+            send_buffer = HK_ADC_REG_PTR_CONFIG;
+            break;
+
+        case HK_ADC_REG_LO_THRESH:
+            send_buffer = HK_ADC_REG_PTR_LOW_THRESH;
+            break;
+
+        case HK_ADC_REG_HI_THRESH:
+            send_buffer = HK_ADC_REG_PTR_HI_THRESH;
+            break;
+
+//        default:
+            //no other register possible
+    }
+
+    // Write to Address Pointer Register - pointing to 'reg'
+    MSS_I2C_write(&g_mss_i2c0, HK_ADC_ADDRESS_GND, &send_buffer, TX_LENGTH, MSS_I2C_RELEASE_BUS);
+
+    // Wait for completion and record the outcome
+    status = MSS_I2C_wait_complete(&g_mss_i2c0, MSS_I2C_NO_TIMEOUT);
+
+    // Proceed only if it is a success
+    if (status == MSS_I2C_SUCCESS)
+    {
+        // Read data from target slave (we always read 2B)
+        MSS_I2C_read(&g_mss_i2c0, HK_ADC_ADDRESS_GND, &rx_buffer[0], RX_LENGTH, MSS_I2C_RELEASE_BUS);
+
+        status = MSS_I2C_wait_complete(&g_mss_i2c0, MSS_I2C_NO_TIMEOUT);
+
+        if(status == MSS_I2C_SUCCESS)
+        {
+            read_value = rx_buffer[0] << 8 | rx_buffer[1];
+            *read_buffer = read_value;
+            err = HK_ADC_NO_ERR;
+            return err;
+        }
+    }
+
+    err = HK_ADC_ERR_READ_FAILED;
+    return err;
+}
+
+
+
+
+/**
+ * @brief HK ADC average voltage calculation function
+ *
+ * This local function calculates the average battery voltage from raw ADC voltage
+ * values. Number of samples to be averaged is defined by NUM_SAMPLES_TO_AVG.
+ *
+ * @param None.
+ *
+ * @return Average battery voltage raw value.
+ *
+ */
+uint16_t hk_adc_calc_avg_voltage(void)
+{
+    uint16_t sum = 0;
+    uint16_t average = 0;
+    for (int i = 0; i < NUM_SAMPLES_TO_AVG; i ++)
```

Comment:
Suggest we "hard-code" the number of samples to average here, to perhaps 8?

Reply:
Done